### ISSUE TITLE
Getting static typing on custom QueryContext properties

### ISSUE THREAD

**Issue by jtlapp** [Vincit/objection.js#668]
Barring a Typescript trick that I'm unaware of, the [typedef for QueryContext](https://github.com/Vincit/objection.js/blob/master/typings/objection/index.d.ts#L154) doesn't allow apps to benefit from static type checking on custom app properties.

I'd rather be forced to reopen the interface to add the properties I need. Maybe there's a good reason it's done this way and not some other?

On the other hand, after removing the `[key: string]: any` from the typedef, I can't figure out *how* to reopen QueryContext for enhancement. The following don't work:

```ts
export interface QueryContext {
  session: Session;
}
```

```ts
declare namespace QueryContext {
  let session: Session;
}
```

I'm still wrapping my head around Typescript typedefs, so my not being able to figure out doesn't tell me it's not possible. If it's not possible, I guess that's why things are as they are. Thanks!

cc @mceachen

**Comment by jtlapp** [comment]
I figured out how to get app-specific static type checking on the Validator context, but it required taking all the Objection types out of the the `Objection` namespace and exporting them directly. See [discussion here](https://github.com/Vincit/objection.js/pull/666#discussion_r158608577). The technique should work with QueryContext as well.

**Comment by jtlapp** [comment]
I now have this working with the typings as extended for QueryBuilder in [temp-qb](https://github.com/jtlapp/temp-qb). I opened up QueryContext as follows:

```ts
declare module 'objection' {
    export interface QueryContext {
        session: Session;
    }
}
```

It's not ideal, as static type checking only occurs when the dev remembers to include this extension and when the dev happens to use the right property name, but it's better than nothing.

@mceachen, you'll want to update to the the latest `temp-qb`, as adding this type check revealed a type inconsistency in my declarations! I've also updated the typings in `temp-qb` to reflect the current thinking on Validator types.
